sklearn/tree/_tree.pxd

```diff
@@ -43,6 +43,67 @@ cdef struct ParentInfo:
     float64_t impurity              # the impurity of the parent
     intp_t n_constant_features      # the number of constant features found in parent
 
+# A record on the stack for depth-first tree growing
+cdef struct StackRecord:
+    intp_t start
+    intp_t end
+    intp_t depth
+    intp_t parent
+    bint is_left
+    float64_t impurity
+    intp_t n_constant_features
+    float64_t lower_bound
+    float64_t upper_bound
+
+cdef extern from "<stack>" namespace "std" nogil:
+    cdef cppclass stack[T]:
+        ctypedef T value_type
+        stack() except +
+        bint empty()
+        void pop()
+        void push(T&) except +  # Raise c++ exception for bad_alloc -> MemoryError
+        T& top()
+
+cdef struct BuildEnv:
```

Comment:
Here as opposed to _tree.pyx? I put BuildEnv here because it will likely need to be visible to event handlers slated for addition, and _tree.pxd seemed like the place to put the "interface" to the module. It can go anywhere that can be made available to external event handlers.

Reply:
Sorry I meant, why do we need the `BuildEnv` in the first place.

Reply:
Or do you mean why is the struct required at all? It _is_ just to encapsulate the function state; we could just as easily pass all the variables/references in as function args, but this seemed cleaner and definitely more expedient during development. More importantly, it's a pattern that will be required for the forthcoming tree build event handling, and might need to be added to the splitter injection.

In general we have this algorithm which remains broadly the same shape but with an arbitrary (and growing) number of variations we'd like to add without foreknowledge and without perpetual updates to the algorithm code itself. Those future additions will require differing degrees of visibility into the algorithm state. So IMO it seemed cleanest (and most performant) to simply encapsulate the algorithm state in a struct whose address we can pass around.